core/src/core/transaction.rs

```diff
@@ -511,52 +511,44 @@ impl TransactionBody {
 			// If we are verifying sort order then verify and
 			// return an error if not sorted lexicographically.
 			body.verify_sorted()?;
-			Ok(body)
 		} else {
 			// If we are not verifying sort order then sort in place and return.
-			let mut body = body;
 			body.sort();
-			Ok(body)
 		}
+		Ok(body)
 	}
 
 	/// Builds a new body with the provided inputs added. Existing
 	/// inputs, if any, are kept intact.
 	/// Sort order is maintained.
-	pub fn with_input(self, input: Input) -> TransactionBody {
-		let mut new_ins = self.inputs;
-		new_ins.push(input);
-		new_ins.sort();
-		TransactionBody {
-			inputs: new_ins,
-			..self
-		}
+	pub fn with_input(mut self, input: Input) -> TransactionBody {
+		self.inputs
+			.binary_search(&input)
+			.err()
+			.map(|e| self.inputs.insert(e, input));
```

Comment:
I'm really torn here - it feels _wrong_ to use the err() here as control flow, but it does seem to be a pattern used (and recommended) in Rust...

`Result<usize, usize>` looks and feels like a nasty hack to me...

One suggestion for readability - can we use `idx` instead of `e`? Its technically an "error" but in practice this is really just the idx to insert at.

Reply:
Yeah! I agree 100% -- one of my least favorite things in the rust API.

Haskell's result type is the value-neutral 'Either'
http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Either.html where Left and Right don't really convey anything.

In Rust, we have a hatred/annoyance for the Right Err type, and using it in a normal control flow feels gross. Rust also used to have Either for this purpose https://github.com/rust-lang/rust/pull/11149

In this case, this is the exact intended use of binary_search as per the docs:

```
If the value is found then Result::Ok is returned, containing the index of the matching element.
If there are multiple matches, then any one of the matches could be returned.
If the value is not found then Result::Err is returned,
containing the index where a matching element could be inserted while maintaining sorted order.
```

We could also use `map_err` if you prefer, but I prefer to use `err()` because it actively discards the Ok result, which feels safer in general (`map_err` leaves it untouched which might cause you to later forget that the `err()` was the desired type)